docs/notebooks/autodiff_remat.md

```diff
@@ -0,0 +1,548 @@
+    f1, f2 = funs
+    return lambda x: f1(f2(x))
+  else:
+    f1 = recursive_checkpoint(funs[:len(funs)//2])
+    f2 = recursive_checkpoint(funs[len(funs)//2:])
+    return lambda x: f1(jax.checkpoint(f2)(x))
+```
+
+```{code-cell}
+f = recursive_checkpoint([jnp.sin] * 8)
+print_saved_residuals(f, 3.)
+```
+
+```{code-cell}
+f = recursive_checkpoint([jnp.sin] * 16)
+print_saved_residuals(f, 3.)
+```
+
++++ {"id": "D0yhcndHX0yN"}
+
+The cost here, as usual, is recomputation: in particular, we end up performing $\mathcal{O}(\log_2 D)$ times as many FLOPs:
+
+```{code-cell}
+f = chain_compose([jnp.sin] * 8)
+print_fwd_bwd(f, 3.)
+```
+
+```{code-cell}
+f = recursive_checkpoint([jnp.sin] * 8)
+print_fwd_bwd(f, 3.)
+```
+
++++ {"id": "wvSm1yu0yUtd"}
+
+### Practical notes
+
++++ {"id": "LSADkBOCyX9X"}
+
+When differentiated functions are staged out to XLA for compilation, for example by applying `jax.jit` to a function which contains a `jax.grad` call, XLA will automatically optimize the computation, including decisions about when to compute or rematerialize values. As a result, **`jax.checkpoint` often isn't needed for differentiated functions under a `jax.jit`**. XLA will optimize things for you.
+
+One exception is when using staged-out control flow, like `jax.lax.scan`. Automatic compiler optimizations between multiple control flow primitives, e.g. between a forward-pass `scan` and the corresponding backward-pass `scan`, aren't aren't as thorough. As a result, it's often a good idea to use `jax.checkpoint` on the body function we pass to `jax.lax.scan`.
+
+For example, one common pattern in large Transformer models is to express the architecture as a `jax.lax.scan` over layers so as to reduce compilation times. That is, using a simple fully-connected network as an analogy, instead of writing something like this:
+
++++ {"id": "BUeqKFRS5yPU"}
+
+```python
+from typing import Tuple, List
+
+LayerParam = Tuple[jnp.ndarray, jnp.ndarray]  # weights, bias pair for a layer
+ParamsList = List[LayerParam]
+
+def net(params: ParamsList, x: jnp.ndarray):
+  for W, b in params:
+    x = jnp.maximum(jnp.dot(x, W) + b, 0.)
+  return x
+```
+
++++ {"id": "38xDcRwW518P"}
+
+We would instead use `jax.lax.scan`:
+
++++ {"id": "ZU2fwYoG6A4z"}
+
+```python
+StackedWeights = jnp.ndarray  # all weight matrices stacked together
+StackedBiases = jnp.ndarray   # all bias vectors stacked together
+
+all_weights = jnp.stack([W for W, _ in params])
+all_biases = jnp.stack([b for _, b in params])
+
+def layer(x, W_b_pair):
+  W, b = W_b_pair
+  out = jnp.maximum(jnp.dot(x, W) + b, 0.)
+  return out, None
+
+def net(all_weights, all_biases, x):
+  x, _ = jax.lax.scan(layer, x, (all_weights, all_biases))
+  return x
+```
+
++++ {"id": "A-NbCn8A6TFK"}
+
+This scan-over-layers version reduces compile times, but by foiling some XLA optimizations it can lead to inefficient computation of gradients. To mitigate the issue, we would use `jax.checkpoint` on the scanned function:
```

Comment:
```suggestion
This scan-over-layers version reduces compile times, but by foiling some compiler optimizations it can lead to inefficient computation of gradients. To mitigate the issue, we would use `jax.checkpoint` on the scanned function:
```

Reply:
Done!